README
cssed findinfiles plugin
------------------------

This plugin provides a search panel
to be added to cssed, which allows 
you to search recursively for a given
term inside files from a given directory.

Read the COPYING file to know license and
distribution terms.

cssed findinfiles plugin (c) Iago Rubio, 2004-2006



NEWS
NEWS file for ccsed-findinfiles plugin

Version 0.2
 Handles now spaces in path
 Freeze the editor on Mac OS X while searching to avoid crash when clicking on a line



src/callbacks.h
 /*  find in files plugin for cssed (c) Iago Rubio 2004
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Library General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#include <gtk/gtk.h>

void
on_button_fif_search_clicked           (GtkButton       *button,
                                        gpointer         user_data);

void
on_button_fif_set_basedir_clicked      (GtkButton       *button,
                                        gpointer         user_data);

void
on_button_fif_set_basedir_from_current_clicked (GtkButton       *button,
                                        gpointer         user_data);
void
on_button_fif_cancel_search_clicked    (GtkButton       *button,
                                        gpointer         user_data);

void
on_button_fif_clear_clicked           (GtkButton       *button,
                                        gpointer         user_data);

void
on_treeview_fif_output_row_activated   (GtkTreeView     *treeview,
                                        GtkTreePath     *path,
                                        GtkTreeViewColumn *column,
                                        gpointer         user_data);

void
on_popmenu_in_current_activate         (GtkMenuItem     *menuitem,
                                        gpointer         user_data);

void
on_popmenu_in_filedir_activate         (GtkMenuItem     *menuitem,
                                        gpointer         user_data);




src/interface.c
 /*  find in files plugin for cssed (c) Iago Rubio 2004
  *
  *  This program is free software; you can redistribute it and/or modify
  *  it under the terms of the GNU General Public License as published by
  *  the Free Software Foundation; either version 2 of the License, or
  *  (at your option) any later version.
  *
  *  This program is distributed in the hope that it will be useful,
  *  but WITHOUT ANY WARRANTY; without even the implied warranty of
  *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  *  GNU Library General Public License for more details.
  *
  *  You should have received a copy of the GNU General Public License
  *  along with this program; if not, write to the Free Software
  *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
  */

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <string.h>
#include <stdio.h>

#include <gdk/gdkkeysyms.h>
#include <gtk/gtk.h>
#include <plugin.h>

#include "findinfiles.h"
#include "callbacks.h"
#include "interface.h"

#define GLADE_HOOKUP_OBJECT(component,widget,name) \
  g_object_set_data_full (G_OBJECT (component), name, \
    gtk_widget_ref (widget), (GDestroyNotify) gtk_widget_unref)

#define GLADE_HOOKUP_OBJECT_NO_REF(component,widget,name) \
  g_object_set_data (G_OBJECT (component), name, widget)

GtkWidget*
create_findinfiles_ui (CssedFifPlugin* fif)
{
	GtkWidget *findinfiles_ui;
	GtkWidget *vbox_fif;
	GtkWidget *hbox_controls;
	GtkWidget *hbox_controls_2;
	GtkWidget *label_searchterm;
	GtkWidget *entry_search_term;
	GtkWidget *label_basedir;
	GtkWidget *entry_base_dir;
	GtkWidget *combo_base_dir;
	GtkWidget *button_fif_search;
	GtkWidget *checkbutton_depth;
	GtkObject *spinbutton_depth_adj;
	GtkWidget *spinbutton_depth;
	GtkWidget *button_fif_set_basedir;
	GtkWidget *button_fif_set_basedir_from_current;	
	GtkWidget *button_fif_cancel_search;
	GtkWidget *button_fif_clear;
	GtkWidget *scrolledwindow_output;
	GtkWidget *treeview_fif_output;
	GtkTooltips *tooltips;
	GtkCellRenderer *renderer;
	GtkListStore *store;
	GtkTreeViewColumn *column;
	
	tooltips = gtk_tooltips_new ();
	
	findinfiles_ui = gtk_frame_new( NULL );

	// first row
	vbox_fif = gtk_vbox_new (FALSE, 0);
	gtk_widget_show (vbox_fif);
	gtk_container_add (GTK_CONTAINER (findinfiles_ui), vbox_fif);
	
	hbox_controls = gtk_hbox_new (FALSE, 0);
	gtk_widget_show (hbox_controls);
	gtk_box_pack_start (GTK_BOX (vbox_fif), hbox_controls, FALSE, TRUE, 0);
	
	label_searchterm = gtk_label_new (_("Search term"));
	gtk_widget_show (label_searchterm);
	gtk_box_pack_start (GTK_BOX (hbox_controls), label_searchterm, FALSE, FALSE, 5);
	
	entry_search_term = gtk_entry_new ();
	gtk_widget_show (entry_search_term);
	gtk_box_pack_start (GTK_BOX (hbox_controls), entry_search_term, TRUE, TRUE, 5);
	
	checkbutton_depth = gtk_check_button_new_with_mnemonic (_("Depth"));
	gtk_widget_show (checkbutton_depth);
	gtk_box_pack_start (GTK_BOX (hbox_controls), checkbutton_depth, FALSE, FALSE, 5);

	spinbutton_depth_adj = gtk_adjustment_new (1, 1, 100, 1, 10, 10);
	spinbutton_depth = gtk_spin_button_new (GTK_ADJUSTMENT (spinbutton_depth_adj), 1, 0);
	gtk_widget_show (spinbutton_depth);
	gtk_box_pack_start (GTK_BOX (hbox_controls), spinbutton_depth, FALSE, FALSE, 10);

	button_fif_search = gtk_button_new_from_stock ("gtk-find");
	gtk_widget_show (button_fif_search);
	gtk_box_pack_start (GTK_BOX (hbox_controls), button_fif_search, FALSE, FALSE, 0);
	gtk_tooltips_set_tip (tooltips, button_fif_search, _("Click here to start a search"), NULL);
	
	button_fif_cancel_search = gtk_button_new_from_stock ("gtk-cancel");
	gtk_widget_show (button_fif_cancel_search);
	gtk_box_pack_start (GTK_BOX (hbox_controls), button_fif_cancel_search, FALSE, FALSE, 5);
	gtk_widget_set_sensitive (button_fif_cancel_search, FALSE);	

	// second row
	hbox_controls_2 = gtk_hbox_new (FALSE, 0);
	gtk_widget_show (hbox_controls_2);
	gtk_box_pack_start (GTK_BOX (vbox_fif), hbox_controls_2, FALSE, TRUE, 0);

	label_basedir = gtk_label_new (_("Base directory"));
	gtk_widget_show (label_basedir);
	gtk_box_pack_start (GTK_BOX (hbox_controls_2), label_basedir, FALSE, FALSE, 5);

	button_fif_set_basedir_from_current = gtk_button_new_with_label (_("From document"));
	gtk_widget_show (button_fif_set_basedir_from_current);
	gtk_box_pack_start (GTK_BOX (hbox_controls_2), button_fif_set_basedir_from_current, FALSE, FALSE, 0);
	gtk_tooltips_set_tip (tooltips, button_fif_set_basedir_from_current, _("Click here to set the base directory from current's document one."), NULL);
	
	button_fif_set_basedir = gtk_button_new_from_stock ("gtk-open");
	gtk_widget_show (button_fif_set_basedir);
	gtk_box_pack_start (GTK_BOX (hbox_controls_2), button_fif_set_basedir, FALSE, FALSE, 0);
	gtk_tooltips_set_tip (tooltips, button_fif_set_basedir, _("Click here to set the base directory"), NULL);

	entry_base_dir = gtk_entry_new ();
	gtk_widget_show (entry_base_dir);
	gtk_box_pack_start (GTK_BOX (hbox_controls_2), entry_base_dir, TRUE, TRUE, 5);
	gtk_tooltips_set_tip (tooltips, entry_base_dir, _("Enter a base directory to search"), NULL);
	
	button_fif_clear = gtk_button_new_from_stock ("gtk-clear");
	gtk_widget_show (button_fif_clear);
	gtk_box_pack_start (GTK_BOX (hbox_controls_2), button_fif_clear, FALSE, FALSE, 5);

	// /////////
	scrolledwindow_output = gtk_scrolled_window_new (NULL, NULL);
	gtk_widget_show (scrolledwindow_output);
	gtk_box_pack_start (GTK_BOX (vbox_fif), scrolledwindow_output, TRUE, TRUE, 5);
	
	treeview_fif_output = gtk_tree_view_new ();
	gtk_widget_show (treeview_fif_output);
	gtk_container_add (GTK_CONTAINER (scrolledwindow_output), treeview_fif_output);
	
	renderer = gtk_cell_renderer_text_new ();	
	column =	gtk_tree_view_column_new_with_attributes ("",  renderer,  "text", 0, NULL);
	gtk_tree_view_column_set_resizable(GTK_TREE_VIEW_COLUMN(column),   TRUE);
	gtk_tree_view_insert_column (GTK_TREE_VIEW (treeview_fif_output),	column, 0);
	store =	gtk_list_store_new (1, G_TYPE_STRING);	
	gtk_tree_view_set_model (GTK_TREE_VIEW (treeview_fif_output),	 GTK_TREE_MODEL (store));
	g_object_unref(store);
	gtk_tree_view_set_headers_visible (GTK_TREE_VIEW (treeview_fif_output), FALSE);
	
	fif->basedir_entry = entry_base_dir;
	fif->searchterm_entry = entry_search_term;
	fif->treeview = treeview_fif_output;
	fif->checkbox_depth = checkbutton_depth;
	fif->spinbutton_depth = spinbutton_depth;
	fif->button_search = button_fif_search;
	fif-> button_cancel = button_fif_cancel_search;
	
	g_signal_connect ((gpointer) button_fif_search, "clicked",
					G_CALLBACK (on_button_fif_search_clicked),
					fif);
	g_signal_connect ((gpointer) button_fif_set_basedir, "clicked",
					G_CALLBACK (on_button_fif_set_basedir_clicked),
					fif);
	g_signal_connect ((gpointer) button_fif_set_basedir_from_current, "clicked",
					G_CALLBACK (on_button_fif_set_basedir_from_current_clicked),
					fif);
	g_signal_connect ((gpointer) button_fif_cancel_search, "clicked",
					G_CALLBACK (on_button_fif_cancel_search_clicked),
					fif);
	g_signal_connect ((gpointer) button_fif_clear, "clicked",
					G_CALLBACK (on_button_fif_clear_clicked),
					fif);
	g_signal_connect ((gpointer) treeview_fif_output, "row_activated",
					G_CALLBACK (on_treeview_fif_output_row_activated), fif);

	return findinfiles_ui;
}


void
create_pop_menu_entry (CssedFifPlugin* fif)
{

	GtkWidget* cssed_popmenu;
	GtkWidget* image;
	GtkWidget* topmenu_item;
	GtkWidget* topmenu_menu;
	GtkWidget* in_current;
	GtkWidget* in_this_files_dir;

	cssed_popmenu = cssed_plugin_get_pop_menu( CSSED_PLUGIN(fif) );

	image = gtk_image_new_from_stock ("gtk-find", GTK_ICON_SIZE_MENU);
	gtk_widget_show (image);

	topmenu_item = gtk_image_menu_item_new_with_label (_("Find in files"));
	gtk_image_menu_item_set_image (GTK_IMAGE_MENU_ITEM (topmenu_item), image);
	gtk_widget_show (topmenu_item);
	gtk_container_add (GTK_CONTAINER (cssed_popmenu), topmenu_item);
	
	topmenu_menu = gtk_menu_new ();
	gtk_menu_item_set_submenu (GTK_MENU_ITEM (topmenu_item), topmenu_menu);
	                                           
	in_current = gtk_image_menu_item_new_with_label (_("In current base directory"));
	gtk_widget_show (in_current);
	gtk_container_add (GTK_CONTAINER (topmenu_menu), in_current);
	
	in_this_files_dir = gtk_image_menu_item_new_with_label (_("In file's base directory"));
	gtk_widget_show (in_this_files_dir);
	gtk_container_add (GTK_CONTAINER (topmenu_menu), in_this_files_dir);

	fif->popmenu_item = topmenu_item;	

	g_signal_connect ((gpointer) in_current, "activate", G_CALLBACK (on_popmenu_in_current_activate), fif);
	g_signal_connect ((gpointer) in_this_files_dir, "activate", G_CALLBACK (on_popmenu_in_filedir_activate), fif);
}


src/interface.h
 /*  find in files plugin for cssed (c) Iago Rubio 2004
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Library General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

GtkWidget* create_findinfiles_ui (CssedFifPlugin* fif);
void create_pop_menu_entry (CssedFifPlugin* fif);



src/findinfiles.h
 /*  find in files plugin for cssed (c) Iago Rubio 2004
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Library General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

/*
 * Standard gettext macros.
 */
#ifdef ENABLE_NLS
#  include <libintl.h>
#  undef _
#  define _(String) dgettext (PACKAGE, String)
#  ifdef gettext_noop
#    define N_(String) gettext_noop (String)
#  else
#    define N_(String) (String)
#  endif
#else
#  define textdomain(String) (String)
#  define gettext(String) (String)
#  define dgettext(Domain,Message) (Message)
#  define dcgettext(Domain,Message,Type) (Message)
#  define bindtextdomain(Domain,Directory) (Domain)
#  define _(String) (String)
#  define N_(String) (String)
#endif

#define FIF_PLUGIN(n) (CssedFifPlugin*) n
#define CSSED_PLUGIN(obj) (obj)->plugin

enum {
	FIF_THREAD_INIT_STATE,
	FIF_THREAD_CONTINUE,
	FIF_THREAD_STOP
};

typedef struct _CssedFifPlugin {
	CssedPlugin* plugin;
	GtkWidget* user_interface;
	GtkWidget* popmenu_item;
	GtkWidget* basedir_entry;
	GtkWidget* searchterm_entry;
	GtkWidget* checkbox_depth;
	GtkWidget* spinbutton_depth;
	GtkWidget* treeview;
	GtkWidget* button_search;
	GtkWidget* button_cancel;
	//GMutex* mutex;
	guint16 flag;
	gchar* command;
} CssedFifPlugin;







src/callbacks.c
 /*  find in files plugin for cssed (c) Iago Rubio 2004
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Library General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#if defined(BSD) || defined(DARWIN) 
 #include <errno.h>
#endif

#include <gtk/gtk.h>
#include <plugin.h>

#include "findinfiles.h"
#include "callbacks.h"
#include "interface.h"

// Thread func
gpointer fif_search_proc (gpointer data);
// utility func
gboolean is_string_all_digits(gchar* string);

typedef struct _DocLineData {
	CssedPlugin *plugin;
	gint line;
} DocLineData;

#if defined(BSD) || defined(DARWIN)
int
getline(char** line, size_t* size, FILE* fp)
{
	size_t i;
	static const size_t grow_rate = 80; 
	int c;

	if (line == NULL || size == NULL || fp == NULL) { 
		errno = EINVAL;
		return -1;
	}
	if (*line == NULL ) { 
		*line = g_malloc0(sizeof(gchar) * grow_rate);
		*size = grow_rate;
	}

	i = 0;
	while (1) {
		c = fgetc(fp);
		if (c == EOF)
			break;
		if ((*size - i) <= 1) {
			*size += grow_rate;
			*line = (char*)g_realloc(*line, *size);
			if (*line == NULL) {
				errno = ENOMEM;
				return -1;
			}
		}
		*(*line + i++) = (char)c;
		if (c == '\n')
			break;
	}

	*(*line + i) = 0;

	if( ferror(fp) != 0 ){
		return -1;
	}else{
		return i;
	}
}
#endif


// for variable substitution, don't use it with large text as it uses a copy - so it doubles the
// memory needed for the string substitution.
// THIS IS FROM UTILS.C BUT WILL BE ON THE PLUGIN'S INTERFACE - FIXME
gchar* string_replace_all(gchar *string, gchar *replaced_str, gchar *replacement_str, gboolean free_string)
{
	gchar* string_copy; // we'll manipulate it so we need a copy
	gchar *iterator, *match, *endstr; // catch position pointers
	GString *newstring;

	g_return_val_if_fail(string != NULL, NULL);
	g_return_val_if_fail(replaced_str != NULL, NULL);
	g_return_val_if_fail(replacement_str != NULL, NULL);

	newstring = g_string_new("");
	string_copy = g_strdup(string);
	endstr = string_copy + strlen(string_copy);
	iterator = string_copy;

	while( (match = strstr (iterator, replaced_str)) != NULL ){
		*match = '\0';
		newstring = g_string_append(newstring, iterator);
		newstring = g_string_append(newstring, replacement_str);
		iterator = match + strlen(replaced_str);
	}
	
	// if no match it'll copy the whole string, if there's something
	// before last match will append the leading text.
	if( iterator != endstr )
		newstring = g_string_append(newstring, iterator);

	g_free(string_copy);
	string_copy = g_strdup(newstring->str);
	g_string_free(newstring, TRUE);
	
	if( free_string ) g_free(string);
	return string_copy;
} 

void
on_button_fif_search_clicked           (GtkButton       *button,
                                        gpointer         user_data)
{
	GtkWidget* basedir_entry;
	GtkWidget* searchterm_entry;
	GtkWidget* checkbox_depth;
	GtkWidget* spinbutton_depth;
	GError* error = NULL;
	gchar* basedir_str;
	gchar* searchterm_str;
	gchar* depth_str;
	gchar* cmd_line;
	gchar* old_ptr;
	GThread* thread;
	CssedFifPlugin* fif;

	fif = FIF_PLUGIN(user_data);
	basedir_entry = fif->basedir_entry;
	searchterm_entry = fif->searchterm_entry;
	checkbox_depth = fif->checkbox_depth;
	spinbutton_depth = fif->spinbutton_depth;

	basedir_str = gtk_editable_get_chars(GTK_EDITABLE(basedir_entry), 0, -1);
	if( strlen(basedir_str) <= 0 ){
		cssed_plugin_error_message(_("Base directory not set"), _("Please set the base directory where the search will start"));
		g_free( basedir_str );
		return;
	}

	searchterm_str = gtk_editable_get_chars(GTK_EDITABLE(searchterm_entry), 0, -1);
	if( strlen(searchterm_str) <= 0 ){
		cssed_plugin_error_message(_("Search term not set"), _("Please enter a search term to search for."));
		g_free( searchterm_str );
		g_free( basedir_str );
		return;
	}	
	cmd_line = g_strdup_printf ("find \"%s\" ",  basedir_str);
	
	if( gtk_toggle_button_get_active(GTK_TOGGLE_BUTTON(checkbox_depth)) ){
		depth_str = g_strdup_printf( "-maxdepth %d", gtk_spin_button_get_value_as_int(GTK_SPIN_BUTTON(spinbutton_depth)) );
		old_ptr = cmd_line;
		cmd_line = g_strdup_printf("%s %s", old_ptr, depth_str );
		g_free(old_ptr);
		g_free(depth_str);
	}

#if defined(DARWIN) // was reported by Michelle that OsX doesn't work without the -maxdepth parameter.
	if( !(gtk_toggle_button_get_active(GTK_TOGGLE_BUTTON(checkbox_depth))) ){
		depth_str = g_strdup_printf( "-maxdepth 1" );
		old_ptr = cmd_line;
		cmd_line = g_strdup_printf("%s %s", old_ptr, depth_str );
		g_free(old_ptr);
		g_free(depth_str);
	}
#endif

	searchterm_str = string_replace_all (searchterm_str, "\"", "\\\"", TRUE); // clean up quotes	
	searchterm_str = string_replace_all (searchterm_str, "`", "\\`", TRUE); // clean up backticks
	searchterm_str = string_replace_all (searchterm_str, "$", "\\$", TRUE); // clean up $
	searchterm_str = string_replace_all (searchterm_str, "\\", "\\\\", TRUE); // clean up \
	
	old_ptr = cmd_line;
	cmd_line = g_strdup_printf("%s -print0 -type f|xargs -0 grep -n \"%s\"", cmd_line, searchterm_str );
	g_free( old_ptr );	
	g_free( searchterm_str );
	
	if( fif->command ) g_free( fif->command );
	fif->command = cmd_line;
	thread = g_thread_create  (fif_search_proc,fif,FALSE ,&error);
	if( thread == NULL ){
		g_free( fif->command );
		fif->command = NULL;
		cssed_plugin_error_message(_("Search error"), _("Unable to launch the search proccess"));
	}
}


void
on_button_fif_set_basedir_clicked      (GtkButton       *button,
                                        gpointer         user_data)
{
	GtkWidget* basedir;
	gchar* dirname;
	gint response;
	gchar* basedir_str;
	gchar* full_path;
	CssedFifPlugin* fif;

#ifdef GTK_ATLEAST_2_4
	gchar* filename;
#else
	G_CONST_RETURN gchar* filename;
#endif
	fif = FIF_PLUGIN(user_data);
	basedir = fif->basedir_entry;

	basedir_str = gtk_editable_get_chars(GTK_EDITABLE(basedir), 0, -1);

	if( strlen(basedir_str) > 0 ){
		if( !g_str_has_suffix (basedir_str, G_DIR_SEPARATOR_S) ){
			full_path = g_strdup_printf("%s%s", basedir_str, G_DIR_SEPARATOR_S);	
			filename = cssed_plugin_prompt_for_directory_to_open(fif->plugin, _("Select a directory"), full_path);
			g_free( full_path );
		}else{
			filename = cssed_plugin_prompt_for_directory_to_open(fif->plugin, _("Select a directory"), basedir_str);
		}
	}else{
		filename = cssed_plugin_prompt_for_directory_to_open(fif->plugin, _("Select a directory"), NULL);
	}

	g_free( basedir_str );
	
	if( filename != NULL ){
		if( !g_file_test( filename, G_FILE_TEST_IS_DIR) ){			
			dirname = g_path_get_dirname ( filename );
			gtk_entry_set_text( GTK_ENTRY(basedir),  dirname);
			g_free( dirname );
		}else{
			gtk_entry_set_text( GTK_ENTRY(basedir),  filename);			
		}
		g_free(filename);
	}
}


void
on_button_fif_set_basedir_from_current_clicked  (GtkButton       *button,
                                        gpointer         user_data)
{
	GtkWidget *basedir;
	gchar *dirname, *filename;
	CssedFifPlugin *fif;

	fif = FIF_PLUGIN(user_data);
	basedir = fif->basedir_entry;
	filename = cssed_plugin_document_get_filename (fif->plugin);
	
	if( filename ){
		dirname = g_path_get_dirname(filename);
		gtk_entry_set_text( GTK_ENTRY(basedir),  dirname);		
		g_free(filename);
		g_free(dirname);
	}else{
		cssed_plugin_error_message(_("File is not saved"), _("The file is not saved on this, cannot use\nthis file's base directory"));
	}	
}

void
on_button_fif_cancel_search_clicked    (GtkButton       *button,
                                        gpointer         user_data)
{
	CssedFifPlugin* fif;
	guint16 flag;
	// No need for a mutex as there'll not be two threads at one time.
	// The running thread will "block" the search button from process
	// start to end.
	fif = FIF_PLUGIN(user_data);
	fif->flag = FIF_THREAD_STOP;
}

void
on_button_fif_clear_clicked           (GtkButton       *button,
                                        gpointer         user_data)
{
	CssedFifPlugin* fif;
	GtkListStore* store;

	fif = FIF_PLUGIN(user_data);
	store = GTK_LIST_STORE(gtk_tree_view_get_model(GTK_TREE_VIEW(fif->treeview)));
	gtk_list_store_clear(store);
}

// Thread func
gpointer
fif_search_proc               (gpointer data)
{
	CssedFifPlugin* fif;
	GtkWidget* search_button;
	GtkWidget* cancel_button;
	GtkWidget *popmenu;
	GtkListStore* store;
	GtkTreeIter iter;
	//GtkTreePath* path;
	//GMutex* mutex;
	gchar* outline;
	guint16 flag;
	gint c;
	FILE* fp;
	char * line = NULL;
#if defined(BSD) || defined(DARWIN)
	size_t len = 0;
	int read;
#else
	size_t len = 0;
	ssize_t read;
#endif

	fif = FIF_PLUGIN(data);	
	//mutex = fif->mutex;
	flag = fif->flag;
	search_button = fif->button_search;
	cancel_button = fif->button_cancel;
	popmenu = fif->popmenu_item;
	store = GTK_LIST_STORE(gtk_tree_view_get_model(GTK_TREE_VIEW(fif->treeview)));

	gdk_threads_enter();
	gtk_widget_set_sensitive(search_button, FALSE); 
	gtk_widget_set_sensitive(cancel_button, TRUE);
	gtk_widget_set_sensitive(popmenu, FALSE);
#if defined(DARWIN)
	gtk_widget_set_sensitive(fif->treeview, FALSE);
#endif
	gdk_threads_leave();

	//g_mutex_lock( mutex );
	fif->flag = FIF_THREAD_CONTINUE;
	//g_mutex_unlock( mutex );

	fp = popen( fif->command, "r");
	if(fp == NULL){
		gdk_threads_enter();
		gtk_widget_set_sensitive(search_button, TRUE);
		gtk_widget_set_sensitive(cancel_button, FALSE); 
#if defined(DARWIN)
		gtk_widget_set_sensitive(fif->treeview, TRUE);
#endif		
		gdk_threads_leave();

		//g_mutex_lock( mutex );
		fif->flag = FIF_THREAD_INIT_STATE;
		//g_mutex_unlock( mutex );	

		if( fif->command ) g_free( fif->command );
		fif->command = NULL;

		gdk_threads_enter();
		cssed_plugin_error_message(_("Search error"), _("Cannot open pipe")); 
		gdk_threads_leave();
		return;
	}else{
		gdk_threads_enter();
		gtk_list_store_clear(store);
		cssed_plugin_select_page_with_widget_in_footer(CSSED_PLUGIN(fif), fif->user_interface);
		gtk_list_store_append(store, &iter);
		gtk_list_store_set( store, &iter, 0, _("** Starting search ..."), -1);
		gdk_threads_leave();
#if defined(BSD) || defined(DARWIN)
		while ((read = getline(&line, &len, fp)) > 0)
#else
		while ((read = getline(&line, &len, fp)) != -1)
#endif
		{
			if( fif->flag != FIF_THREAD_CONTINUE ){
				if(line) g_free(line);
				break;
			}else{
				g_strstrip(line);
				gdk_threads_enter();
				gtk_list_store_append(store, &iter);
				gtk_list_store_set(store, &iter, 0,line, -1);
				gdk_threads_leave();
				if(line) g_free(line);
			}
			line = NULL;
		}
		gdk_threads_enter();
		gtk_list_store_append(store, &iter);
		gtk_list_store_set(store, &iter, 0, _("** Search end."), -1);
		gdk_threads_leave();
		pclose(fp);
	}	

	fif->flag = FIF_THREAD_INIT_STATE;
	g_free(fif->command);
	fif->command = NULL;

	gdk_threads_enter();
	gtk_widget_set_sensitive(search_button, TRUE);
	gtk_widget_set_sensitive(cancel_button, FALSE);
	gtk_widget_set_sensitive(popmenu, TRUE);
#if defined(DARWIN)
	gtk_widget_set_sensitive(fif->treeview, TRUE);
#endif
	gdk_threads_leave();

	g_thread_exit(EXIT_SUCCESS);
}

gboolean    
fif_move_to_line_on_idle (gpointer user_data)
{
	DocLineData* data = (DocLineData*) user_data;
	cssed_plugin_set_arrow_marker_at_line (data->plugin, data->line);
	return FALSE;
}

void
on_treeview_fif_output_row_activated   (GtkTreeView     *treeview,
                                        GtkTreePath     *path,
                                        GtkTreeViewColumn *column,
                                        gpointer         user_data)
{
	DocLineData* data;
	CssedFifPlugin* fif;
	GtkTreeIter iter;
	GtkListStore* store;
	gchar* outline;
	gchar** entries;
	gchar* file;
	gchar* number;
	gint pos;

	fif = FIF_PLUGIN(user_data);
	store = (GtkListStore*) gtk_tree_view_get_model( treeview );

	if (gtk_tree_model_get_iter (GTK_TREE_MODEL(store), &iter, path))
	{
		gtk_tree_model_get (GTK_TREE_MODEL(store), &iter, 0, &outline, -1);
		entries = g_strsplit  (outline,":",3);
		file = *entries;
		if( file == NULL ){
			g_strfreev( entries );
			return;
		}
		number = *(entries+1);
		if( number == NULL ){
			g_strfreev( entries );
			g_free( outline );
			return;		
		}		
		if( !is_string_all_digits(number) ){
			g_strfreev( entries );
			g_free( outline );
			return;	
		}
		if( !g_file_test(file, G_FILE_TEST_IS_REGULAR) ){
			g_strfreev( entries );
			g_free( outline );
			return;		
		}

		if( !cssed_plugin_is_file_opened(CSSED_PLUGIN(fif), file) ){
			cssed_plugin_open_file(CSSED_PLUGIN(fif), file);
			if( cssed_plugin_is_file_opened(CSSED_PLUGIN(fif), file) ){
				data = g_malloc0(sizeof(DocLineData));
				data->plugin = CSSED_PLUGIN(fif);
				data->line = atoi(number);				
				g_idle_add_full (G_PRIORITY_DEFAULT_IDLE,
                                 fif_move_to_line_on_idle,
                                 data, g_free);                                             

			}		
		}else{
			cssed_plugin_set_arrow_marker_at_line(CSSED_PLUGIN(fif), atoi(number));
		}

		g_strfreev( entries );
		g_free( outline );	
	}
}

gboolean
is_string_all_digits(gchar* string)
{
	gint len;
	gint i;

	len = strlen( string );
	
	for(i=0;i<len;i++){
		if( !g_ascii_isdigit(string[i]) ){
			return FALSE;
		}
	}
	return TRUE;
}

void
on_popmenu_in_current_activate         (GtkMenuItem     *menuitem,
                                        gpointer         user_data)
{
	CssedFifPlugin* fif;
	GtkWidget* searchterm_entry;
	gchar* selection;
	gint page;

	fif = FIF_PLUGIN(user_data);	
	searchterm_entry = fif->searchterm_entry;

	selection = cssed_plugin_get_selected_text( CSSED_PLUGIN(fif) );
	if( selection != NULL  ){
		gtk_entry_set_text( GTK_ENTRY(searchterm_entry), selection );
		on_button_fif_search_clicked (NULL, (gpointer) fif );
		g_free(selection);
	}	
}

void
on_popmenu_in_filedir_activate         (GtkMenuItem     *menuitem,
                                        gpointer         user_data)
{	
	CssedFifPlugin* fif;
	GtkWidget* basedir_entry;
	GtkWidget* searchterm_entry;
	gchar* selection;
	gchar* filename;
	G_CONST_RETURN gchar* file_basedir;
	gint page;

	fif = FIF_PLUGIN(user_data);	
	searchterm_entry = fif->searchterm_entry;
	basedir_entry = fif->basedir_entry,
	
	selection = cssed_plugin_get_selected_text(CSSED_PLUGIN(fif));

	if( selection != NULL  ){
		filename = cssed_plugin_document_get_filename(CSSED_PLUGIN(fif));
		if( filename ){
			file_basedir = g_dirname(filename);
			gtk_entry_set_text(GTK_ENTRY(basedir_entry), (gchar*) file_basedir);
			gtk_entry_set_text(GTK_ENTRY(searchterm_entry), selection);
			on_button_fif_search_clicked (NULL, (gpointer) fif);
			g_free(filename);			
		}else{
			cssed_plugin_error_message(_("File not saved"), _("This file is not saved, so there is no base directory for this file.\nPlease save the file and try again."));
		}
		g_free(selection);
	}	
}
